### PR TITLE
feat!: Add start time for continuous vesting accounts

### LINKED ISSUE #341: Add the ability to creating a vesting account that begins vesting in the future
Currently, the vesting module allows for a vesting period to begin in the future if and only if that vesting account is defined in genesis. This is because we are [forcing the start time to be the block time](https://github.com/celestiaorg/cosmos-sdk/blob/64adc6e40ae679003d2f779387e06c4105c99121/x/auth/vesting/msg_server.go#L63). 

We should instead include a start time in the msg, and then use that start time in the state machine.

ref https://github.com/cosmos/cosmos-sdk/issues/4287
specifically this [comment](https://github.com/cosmos/cosmos-sdk/issues/4287#issuecomment-1564423901) confirming our analysis that there are not any currently known side effects of doing this.

### PR DESCRIPTION
**Pull request by liamsi**
Minimal changes required for: https://github.com/celestiaorg/cosmos-sdk/issues/341

Note: 
- ~~I had to regenerate proto files and was surprised that this changed a bunch generated code where the proto source wasn't modified 🤔~~ fixed 
- only modified the existing tests slightly with a start-date in the "future"
- I would also submit this as a PR to upstream 0.46, 0.47, and main
- It would be ideal to have this tested in e2e tests as well
- ~~Ideally, the added command-line flag would also be tested in some fashion~~ (see: [3176583](https://github.com/celestiaorg/cosmos-sdk/pull/342/commits/31765837f78c2b56759680245c272f67c22b19dd) thanks @evan-forbes )

## Description

Closes: #341 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/auth/vesting/client/cli/tx.go [hunk 1]
```diff
@@ -72,14 +73,18 @@ timestamp.`,
...
 
-			msg := types.NewMsgCreateVestingAccount(clientCtx.GetFromAddress(), toAddr, amount, endTime, delayed)
-
+			msg := types.NewMsgCreateVestingAccount(clientCtx.GetFromAddress(), toAddr, amount, startTime, endTime, delayed)
 			return tx.GenerateOrBroadcastTxCLI(clientCtx, cmd.Flags(), msg)
 		},
 	}
 
 	cmd.Flags().Bool(FlagDelayed, false, "Create a delayed vesting account if true")
+	cmd.Flags().Int64(FlagStartTime, 0, "Optional start time (as a UNIX epoch timestamp) for continuous vesting accounts. If 0 (default), the block's time of the block this tx is committed to will be used.")
```
Comment (liamsi) [inline x/auth/vesting/client/cli/tx.go:87]: Comment for reviewer:

It's a bit odd that end_time is passed as a mandatory arg but start-time as an optional flag: 
`create-vesting-account [to_address] [amount] [end_time] --start-time ...`

File: x/auth/vesting/msg_server.go [hunk 2]
```diff
@@ -44,6 +44,19 @@ func (s msgServer) CreateVestingAccount(goCtx context.Context, msg *types.MsgCre
 		return nil, err
 	}
 
+	if msg.EndTime <= 0 {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "invalid end time")
+	}
+
+	if msg.EndTime <= msg.StartTime {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "invalid start and end time (must be start < end)")
+	}
```
Comment (evan-forbes) [inline x/auth/vesting/msg_server.go:53]: since these are stateless checks, ideally we put these in the validate basic function

File: x/distribution/types/distribution.pb.go [hunk 3]
```diff
@@ -8,7 +8,7 @@ import (
 	_ "github.com/cosmos/cosmos-proto"
 	github_com_cosmos_cosmos_sdk_types "github.com/cosmos/cosmos-sdk/types"
 	types "github.com/cosmos/cosmos-sdk/types"
-	_ "github.com/gogo/protobuf/gogoproto"
+	_ "github.com/cosmos/gogoproto/gogoproto"
```
Comment (liamsi) [inline x/distribution/types/distribution.pb.go:11]: Question: Is this expected to change? Seems weird to me. I used ` make proto-gen`.

Comment (julienrbrt) [reply x/distribution/types/distribution.pb.go:11]: Which version of proto-builder are you using?

Comment (julienrbrt) [reply x/distribution/types/distribution.pb.go:11]: It usually means you are using the wrong version, latest uses the latest `gogo.proto`

https://buf.build/cosmos/gogo-proto/file/5e5b9fdd01804356895f8f79a6f1ddc1:gogoproto/gogo.proto#L36 
vs
https://buf.build/cosmos/gogo-proto/file/bee5511075b7499da6178d9e4aaa628b:gogoproto/gogo.proto#L36

File: x/auth/vesting/client/testutil/suite.go [hunk 4]
```diff
@@ -63,6 +63,21 @@ func (s *IntegrationTestSuite) TestNewMsgCreateVestingAccountCmd() {
...
+				fmt.Sprintf("--%s=%d", cli.FlagStartTime, 4070808800),
+				fmt.Sprintf("--%s=%s", flags.FlagFrom, val.Address),
+				fmt.Sprintf("--%s=true", flags.FlagSkipConfirmation),
+				fmt.Sprintf("--%s=%s", flags.FlagBroadcastMode, flags.BroadcastSync),
+				fmt.Sprintf("--%s=%s", flags.FlagFees, sdk.NewCoins(sdk.NewCoin(s.cfg.BondDenom, sdk.NewInt(10))).String()),
+			},
+			expectErr:    false,
+			expectedCode: 0,
+			respType:     &sdk.TxResponse{},
+		},
```
Comment (evan-forbes) [inline x/auth/vesting/client/testutil/suite.go:80]: added a quick test here, but we're not checking that the state really functioned as intented

**Review by evan-forbes — APPROVED** [review]
LGTM

File: x/auth/vesting/msg_server.go (hunk 2 above)
Comment (liamsi) [reply x/auth/vesting/msg_server.go:53]: Yeah, will add these there instead.

**Comment by liamsi** [comment]
looking into this now: https://github.com/celestiaorg/cosmos-sdk/pull/342#discussion_r1311495059

File: x/distribution/types/distribution.pb.go (hunk 3 above)
Comment (liamsi) [reply x/distribution/types/distribution.pb.go:11]: > Which version of proto-builder are you using?

Note that I only used make `proto-gen` which pulls docker images as you know. 

If I prune all images locally and start from scratch, it starts with: 
```
Generating Protobuf files
Unable to find image 'tendermintdev/sdk-proto-gen:v0.7' locally
v0.7: Pulling from tendermintdev/sdk-proto-gen
```

Comment (liamsi) [reply x/distribution/types/distribution.pb.go:11]: Note that this is the same version as in the latest upstream SDK 0.46.15: https://github.com/cosmos/cosmos-sdk/blob/d6d929843bbd331b885467475bcb3050788e30ca/Makefile#L383-L384

Comment (liamsi) [reply x/distribution/types/distribution.pb.go:11]: The only diff here is that our fork's go.mod differs here:
https://github.com/celestiaorg/cosmos-sdk/blob/64adc6e40ae679003d2f779387e06c4105c99121/go.mod#L20
this dep doesn’t exist in upstream:
https://github.com/cosmos/cosmos-sdk/blob/d6d929843bbd331b885467475bcb3050788e30ca/go.mod#L15-L25

same for 0.46.14 btw:
https://github.com/cosmos/cosmos-sdk/blob/v0.46.14/go.mod

Comment (liamsi) [reply x/distribution/types/distribution.pb.go:11]: @julienrbrt is there anything I should be doing differently?

Comment (liamsi) [reply x/distribution/types/distribution.pb.go:11]: @evan-forbes @rootulp @cmwaters any ideas about the above diff?

File: x/auth/vesting/msg_server.go (hunk 2 above)
Comment (liamsi) [reply x/auth/vesting/msg_server.go:53]: Note that these are already checkec in the corresponding ValidateBasic: https://github.com/celestiaorg/cosmos-sdk/blob/55054282d2df794d9a5fe2599ea25473379ebc3d/x/auth/vesting/types/vesting_account.go#L276-L282

File: x/distribution/types/distribution.pb.go (hunk 3 above)
Comment (julienrbrt) [reply x/distribution/types/distribution.pb.go:11]: > @julienrbrt is there anything I should be doing differently?

No, if it uses the right proto-builder image (with the right gogo.proto), then it should not be adding a cosmos/gogoproto dependency, so I am not sure here.

File: x/auth/vesting/msg_server.go [hunk 5]
```diff
@@ -44,6 +44,19 @@ func (s msgServer) CreateVestingAccount(goCtx context.Context, msg *types.MsgCre
...
+
+	ctx = sdk.UnwrapSDKContext(goCtx)
+	if err := s.BankKeeper.IsSendEnabledCoins(ctx, msg.Amount...); err != nil {
+		return nil, err
+	}
+
```
Comment (rootulp) [inline x/auth/vesting/msg_server.go:59]: [question] why is this check needed? It appears unrelated.

File: x/auth/vesting/types/vesting_account_test.go [hunk 6]
```diff
@@ -62,43 +62,87 @@ func TestGetVestedCoinsContVestingAcc(t *testing.T) {
...
-	vestingCoins = cva.GetVestingCoins(now.Add(12 * time.Hour))
+	vestingCoins = cva.GetVestingCoins(startTime.Add(12 * time.Hour))
 	require.Equal(t, sdk.Coins{sdk.NewInt64Coin(feeDenom, 500), sdk.NewInt64Coin(stakeDenom, 50)}, vestingCoins)
+
+	// require 25% of coins vesting
+	vestingCoins = cva.GetVestingCoins(startTime.Add(18 * time.Hour))
+	require.Equal(t, sdk.Coins{sdk.NewInt64Coin(feeDenom, 250), sdk.NewInt64Coin(stakeDenom, 25)}, vestingCoins)
+}
+
+func TestGetVestingCoinsContVestingAccStartTimeInFuture(t *testing.T) {
```
Comment (rootulp) [inline x/auth/vesting/types/vesting_account_test.go:93]: [question] how does this test differ from the test above titled: `TestGetVestingCoinsContVestingAcc` which uses a start time in the future?

**Review by rootulp — COMMENTED** [review]
[nit] the PR title could use "start time" instead of "start date".

I think this line needs to be updated in the [specs](https://github.com/celestiaorg/cosmos-sdk/blob/c09602f4f291e2cc68a0f4a16ace35ef8be8461e/x/auth/spec/07_client.md?plain=1#L411):

> All vesting accouts created will have their start time set by the committed block's time.

**Review by evan-forbes — APPROVED** [review]

File: x/distribution/types/distribution.pb.go (hunk 3 above)
Comment (liamsi) [reply x/distribution/types/distribution.pb.go]: FYI, this fixed it: https://github.com/celestiaorg/cosmos-sdk/pull/343

File: x/auth/vesting/types/vesting_account_test.go (hunk 6 above)
Comment (liamsi) [reply x/auth/vesting/types/vesting_account_test.go]: I could swear I deleted this one and just used the existing. Great catch. Fixed in [1fb01d7](https://github.com/celestiaorg/cosmos-sdk/pull/342/commits/1fb01d7fdede6237c18d33ae2575d438f61f5358)

**Review by evan-forbes — APPROVED** [review]

**Review by rootulp — APPROVED** [review]
LGTM!
